posts/change-data-capture-cloudsql-postgres-with-apache-airflow-1.md

```diff
@@ -0,0 +1,1016 @@
+
+> This is powerful because you can use that hard interval to further refine your watermarks and obtain data created **right up until the DAG starts running**. With that, you now have near-real time change data capture capabilities. Furthermore, the consistency and high-availability allows for SLAs to be set on this data (and in Airflow), something extremely difficult in a streaming solution.
+
+To do this:
+
+1. If your interval is always going to be the same, in `table_1.sql`, you could hardcode your `low_watermark` adjustment to be:
+    
+    ```python
+    {%- set low_watermark =  execution_date -%}
+    {%- set high_watermark =  execution_date.add(hours=1) -%}
+    SELECT * FROM some_schema.table_1
+    WHERE event_timestamp AT TIME ZONE 'UTC' >= '{{ low_watermark }}' AT TIME ZONE 'UTC'
+    AND event_timestamp AT TIME ZONE 'UTC' < '{{ high_watermark }}'
+    ```
+
+    which you would see renders to
+
+    ```bash
+    # ----------------------------------------------------------
+    # property: export_sql
+    # ----------------------------------------------------------
+    SELECT * FROM some_schema.table_1
+    WHERE event_timestamp AT TIME ZONE 'UTC' >= '2021-01-14T13:00:00+00:00' AT TIME ZONE 'UTC'
+    AND event_timestamp AT TIME ZONE 'UTC' < '2021-01-14T14:00:00+00:00'
+    ```
+
+    getting you changes in the previous hour. 
+
+2. Make the following changes to the rest of your `.sql` files.
+
+3. In your DAG definition, adjust to your interval of choice. Suppose you want to make this run every 30 minutes. The `schedule_interval` parameter also accepts a crontab value. So in the DAG definition, that would be `schedule_interval = '* /30 * * * *'`, and your templated SQL would appear:
+
+    ```bash
+    {%- set low_watermark =  execution_date -%}
+    {%- set high_watermark =  execution_date.add(minutes=30) -%}
+    SELECT * FROM some_schema.table_1
+    WHERE event_timestamp AT TIME ZONE 'UTC' >= '{{ low_watermark }}' AT TIME ZONE 'UTC'
+    AND event_timestamp AT TIME ZONE 'UTC' < '{{ high_watermark }}'
+    ```
+
+    In a future post, I will show you how to set these dynamically through the UI as to allow interval changes and full scans when needed.
+
+### Step 13: Intro to Dynamic Configs with Airflow Variables
+
+Airflow Variables are simple key-value fields that can be added through the Airflow UI and parsed in your DAG. They are useful for adding dynamic configuration values to your DAG file. With this, it's extremely important to realize every loop of the scheduler will parse it, making a connection to the database each time. To make this efficient, it's recommended to add a single JSON-serialized variable, then access individual values as items in a dictionary. To convert our hard-coded strings:
```

Comment:
```suggestion
Airflow Variables are simple key-value fields that can be added through the Airflow UI and parsed in your DAG. They are useful for adding dynamic configuration values to your DAG file. With this, it's extremely important to realize every loop of the scheduler will parse it, meaning that every loop of the scheduler makes a connection to the database. To make this efficient, it's recommended to add a single JSON-serialized variable, then access individual values as items in a dictionary. To convert our hard-coded strings:
```

Reply:
@jwitz I changed this to "With this, it's extremely important to realize that on every loop of the scheduler a query is made to the database for each variable."

Reply:
"With this, it's important to note that for every loop of the scheduler, a query is made to the database for each variable"